### ISSUE TITLE
Suggestion: For key:value filters run models change event refiltering only if filtered key changed

### ISSUE THREAD

**Issue by soswow** [jmorrell/backbone-filtered-collection#3]
If I have big collection and filtered collection with some specific key filtered by. Like "project_id" for example. Now I change "visibility" to all models in original collection. Result: I have 500ms lug because each model triggers "change:visibility" and filtered collection recheck each model. But there is enough meta information at that moment not to do this. In this case models should be rechecked only if project_id property changes.

**Comment by jmorrell** [comment]
Okay, so if I follow the issue is something like this:

``` js
var coll = new Backbone.Collection(/* lots o' models */);
var proxy = new FilteredCollection(coll);

proxy.filterBy('project_id', { project_id: "foo" });

coll.invoke('set', { visibility: true }); // triggers change:visibility on every model
```

Changing each parameter triggers the filter to be invoked, but this should be super fast. Are you sure that FilteredCollection is the bottleneck here? Are you touching the DOM on change events maybe? 

I'll do some benchmarking to make sure. How many models are we talking about?

**Comment by soswow** [comment]
Yes. IT wasn't touching the DOM. I will make an example for you. Also I've
made my implementation of this. Will show you as well.

**Comment by soswow** [comment]
Ok. Here is idea of proposed change:
https://github.com/soswow/backbone-filtered-collection/commit/2d6cc8d7bc66bbfed136d0d4a97cb47f076e2686

I tried to make an example. The thing is that it boost performance in my production code, but I can't figured out what is it so special, to make proper example.

I ended up with something like 

``` coffeescript
c1 = new Backbone.Collection()
  for i in [0...20]
    c1.add(new Backbone.Model({a:10, c:20}))

  c2s = []
  for i in [0...20]
    c2 = new Backbone.Collection()
    for i in [0...10]
      c2.add(new Backbone.Model({a:10, c:20}))
    c2s.push c2

  c3s = []
  for i in [0...10]
    c3 = new Backbone.Collection()
    for i in [0...40]
      c3.add(new Backbone.Model({a:10, c:20}))
    c3s.push c3


console.time "Apply filter"
  fc1 = new FilteredCollection(c1)
  fc1.filterBy "a-filter", a: 10

  fc1.each (c1, i) ->
    c1.dependecies = new FilteredCollection(c2s[i])
    c1.dependecies.filterBy "a-filter", a: 10
    c1.dependecies.each (c2, j) ->
      c2.dependecies = new FilteredCollection(c3s[j])
      c2.dependecies.filterBy "a-filter", a: 10
  console.timeEnd "Apply filter"


console.time "Setting b value"
  console.profile("Set b value")
  fc1.each (c1) ->
    c1.set b: 50
    c1.dependecies.each (c2) ->
      c2.set b: 50
      c2.dependecies.each (c3) ->
        c3.set b: 50
  console.profileEnd("Set b value")
  console.timeEnd "Setting b value"
```

It doesn't make so big improvement with this example. 
I can show you Flame chart for my production code with and without "fix"

This is without fix.
![screenshot 2014-07-01 16 31 08](https://cloud.githubusercontent.com/assets/108065/3443546/140e7fbc-0124-11e4-93b6-92112f74b03a.png)

Same peace of code with the fix.
![screenshot 2014-07-01 16 32 21](https://cloud.githubusercontent.com/assets/108065/3443548/2ad89480-0124-11e4-8dda-d58aa1502b28.png)

**Comment by soswow** [comment]
I think this does it so slow because of number of filter collection and they are accessed one from another. Maybe this is the reason. Because when I tried with one long collection - it was the same.

``` coffeescript
workspaces.each (ws) ->
      ws.getProjectsFor("popdown").each (project) ->
        project.set {visible: true}
        project.getTasksFor("popdown").each (task) ->
          task.set {visible: true}
      ws.getClientsFor("popdown").each (client) ->
        client.set {visible: true}
```

here is the code I have. All those `getClientsFor` methods return FilteredCollection. Those collection have one filter each. Something like `filterBy "somename", project_id: @model.id`

**Comment by jmorrell** [comment]
@soswow Sorry for the delay in responding. I'm traveling currently, but I'll try to review this tonight.

**Comment by soswow** [comment]
No problem. Anyway we decided to use backbone-relational in our project.
Anyway it's much easier to use and much closer to goals we have. Also,
using FilteredCollection with 1000 objects and many collection we end up
with 400MB memory taken away. We tried to investigate whats take so much
space. Our suspicion is about internal caching mechanism Filtered
collection has.

Btw, I think my changes might be not good. Question is - if change in other
properties (ones not mentioned in filters) bubble up in filtered
collection? (maybe it's not a problem and this is responsibility of
proxy-collection part of your library.
